python/ray/util/sgd/torch/worker_group.py

```diff
@@ -0,0 +1,480 @@
+                      use_gpu):
+        self.num_workers = num_workers
+        self.params = params
+        self.dist_params = dist_params
+        self.initialization_hook = initialization_hook
+        self.timeout_s = timeout_s
+        self.num_cpus_per_worker = num_cpus_per_worker
+        self.use_gpu = use_gpu
+        if num_workers == 1:
+            RemoteRunner = ray.remote(
+                num_cpus=num_cpus_per_worker,
+                num_gpus=int(use_gpu))(TorchRunner)
+            self.remote_workers = [RemoteRunner.remote(**params)]
+            self.num_workers = num_workers
+            ray.get(self.remote_workers[0].setup_operator.remote())
+        else:
+            self._init_workers(num_workers, {
+                **params,
+                **dist_params
+            }, num_cpus_per_worker, use_gpu)
+            self.num_workers = num_workers
+
+            if initialization_hook:
+                self.apply_all_workers(initialization_hook)
+
+            address = ray.get(self.remote_workers[0].setup_address.remote())
+
+            ray.get(
+                self._setup_process_group(address, timeout_s,
+                                          self.num_workers))
+
+            ray.get(self._setup_operator())
+
+    def _apply_all_operators(self, fn):
+        remote_calls = [
+            w.apply_operator.remote(fn) for w in self.remote_workers
+        ]
+        return remote_calls
+
+    def apply_all_operators(self, fn):
+        return ray.get(self._apply_all_operators(fn))
+
+    def _apply_all_workers(self, fn):
+        return [w.apply.remote(fn) for w in self.remote_workers]
+
+    def apply_all_workers(self, fn):
+        return ray.get(self._apply_all_workers(fn))
+
+    def get_local_operator(self):
+        raise NotImplementedError(
+            "Cannot return a local operators if all "
+            "workers are remote. Set use_local to True in"
+            "TorchTrainer to access a local operator.")
+
+    def get_model(self):
+        ready, _ = ray.wait(
+            [r.get_models.remote() for r in self.remote_workers])
+        models = ray.get(ready[0])
+        return models
+
+    def _load_state_id(self, state_id):
+        remote_calls = [
+            worker.load_state_stream.remote(state_id)
+            for worker in self.remote_workers
+        ]
+        return remote_calls
+
+    def load_state_dict(self, state_dict, blocking=False):
+        _buffer = io.BytesIO()
+        torch.save(state_dict, _buffer)
+        stream = _buffer.getvalue()
+        state_id = ray.put(stream)
+
+        remote_calls = self._load_state_id(state_id)
+
+        if blocking:
+            ray.get(remote_calls)
+
+    def state_dict(self, to_cpu=False):
+        ready, _ = ray.wait(
+            [r.state_dict.remote(to_cpu) for r in self.remote_workers],
+            num_returns=1)
+        return ray.get(ready[0])
+
+    def _train(self, num_steps, profile, info, dataset=None):
```

Comment:
can you leave note about why this is different from below (`train`)?

Reply:
done